### PR TITLE
exim: Correct ZAP Message export/import consistency

### PR DESCRIPTION
**Pull request by kingthorin**
- CHANGELOG > Added change note.
- PopupMenuExportMessages > Ensure the export delimiter format is consistent with the expectation of the Import functionality.
- LogsImporter > Changed visibility of read(File) method to allow testing, added a note covering the same.
- LogsImporterUnitTest > Added tests to ensure import works with various delimiters. Plus associated resource files.

https://github.com/zaproxy/zap-extensions/blob/248b4a0850ba57650ce32032bcd2a7f7458b4b34/addOns/exim/src/main/java/org/zaproxy/addon/exim/log/LogsImporter.java#L181

### CODE REVIEW

**Comment by thc202** [comment]
Be good to add tests.

File: addOns/exim/src/main/java/org/zaproxy/addon/exim/PopupMenuExportMessages.java [hunk 1]
```diff
@@ -115,19 +116,13 @@ private void exportHistory(HttpMessage msg, BufferedOutputStream bos, boolean re
         try {
             if (responsesOnly) {
                 if (!msg.getResponseHeader().isEmpty()) {
-                    bos.write(NEWLINE.getBytes());
```
Comment (thc202) [inline addOns/exim/src/main/java/org/zaproxy/addon/exim/PopupMenuExportMessages.java:118]: This could be kept here.

**Comment by thc202** [comment]
The import could also be tweaked to not require the spaces and allow to import what was already exported.

**Comment by kingthorin** [comment]
> Be good to add tests.

Sure I'll look at that over the weekend. Good with the existing code changes?

**Comment by thc202** [comment]
Yes.

File: addOns/exim/src/main/java/org/zaproxy/addon/exim/log/LogsImporter.java [hunk 2]
```diff
@@ -173,12 +173,18 @@ private boolean processInput(File newFile, LogType logChoice) {
...
+     *
+     * @throws IOException
+     */
+    protected static List<String> readFile(File file) throws IOException {
         List<String> parsed = new ArrayList<>();
         Charset charset = StandardCharsets.US_ASCII;
         BufferedReader reader = Files.newBufferedReader(file.toPath(), charset);
         Scanner sc = new Scanner(reader);
-        sc.useDelimiter(Pattern.compile("====\\s[0-9]*\\s=========="));
+        sc.useDelimiter(Pattern.compile("====?\\s?[0-9]*\\s=========="));
```
Comment (kingthorin) [inline addOns/exim/src/main/java/org/zaproxy/addon/exim/log/LogsImporter.java:187]: I did identify one issue with this implementation (though it always would have existed). There's no way for the scanner to know how many things it should be finding, so if it doesn't find a match it treats the entire content as a single 'blob'. (I guess similarly if there were multiple good delimiters with a bad one in the middle then the 'bad' message will end up "stuck" together with the one ahead of it.)

This shouldn't be a major issue as it should only come up with a user manually creates the file/entries themselves.

File: addOns/exim/src/main/java/org/zaproxy/addon/exim/log/LogsImporter.java [hunk 3]
```diff
@@ -173,12 +173,18 @@ private boolean processInput(File newFile, LogType logChoice) {
...
     }
 
-    private static List<String> readFile(File file) throws IOException {
+    /**
+     * This method is {@code protected} to facilitate Unit Testing. It is not expected to be used
+     * directly.
+     *
+     * @throws IOException
+     */
+    protected static List<String> readFile(File file) throws IOException {
```
Comment (thc202) [inline addOns/exim/src/main/java/org/zaproxy/addon/exim/log/LogsImporter.java:182]: package and the JavaDoc is not needed.

File: addOns/exim/src/test/resources/org/zaproxy/addon/exim/log/110.txt [hunk 4]
```diff
@@ -0,0 +1,3300 @@
+==== 1 ==========
```
Comment (thc202) [inline addOns/exim/src/test/resources/org/zaproxy/addon/exim/log/110.txt:1]: IMHO this should be generated dynamically.

File: addOns/exim/src/test/resources/org/zaproxy/addon/exim/log/fourEqualsNoSpace.txt [hunk 5]
```diff
@@ -0,0 +1,59 @@
+====1 ==========
```
Comment (thc202) [inline addOns/exim/src/test/resources/org/zaproxy/addon/exim/log/fourEqualsNoSpace.txt:1]: IMO these should be trimmed to have only the minimum data required (we are testing the delimiters not the contents of the messages).

File: addOns/exim/src/main/java/org/zaproxy/addon/exim/log/LogsImporter.java (hunk 2 above)
Comment (thc202) [reply addOns/exim/src/main/java/org/zaproxy/addon/exim/log/LogsImporter.java:187]: There's also the (unlikely) case of the delimiters being in the content of the messages. We should look at deprecating this format.

File: addOns/exim/src/test/resources/org/zaproxy/addon/exim/log/fourEqualsNoSpace.txt (hunk 5 above)
Comment (thc202) [reply addOns/exim/src/test/resources/org/zaproxy/addon/exim/log/fourEqualsNoSpace.txt:1]: `User-Agent`, `Accept`, and `Accept-Language` can also be removed.

**Comment by kingthorin** [comment]
Done

**Review by thc202 — APPROVED** [review]

**Review by psiinon — APPROVED** [review]

**Pull request merged**